README.md
# Github deploy status

> Node.js scripts that update a Github repository's deployment status.

Did you know that you keep track of your deployments right on Github.com?
You can do this through the **Environment** tab!

![environments](assets/environment.png)

_What is the **Environment** tab and why don't I see it on my repositories?_

The Environment tab only shows up if you add a **deployment**
to your Github repository. Once you've added a deployment, you can
update its status to show whether it's pending, in progress, successful,
failed, or errored. Some third-party integrations such
[Heroku](https://devcenter.heroku.com/articles/github-integration)
do this for you, so when you deploy through that service, it shows
up on your Github page. If you deploy through custom CI scripts
using services such as Travis CI, this won't show up automatically.

This package provides CLI scripts to help interface with Github's
[deployments API](https://developer.github.com/v3/repos/deployments)
so you can track your deployments right in your pull requests:

![pullrequest](assets/pullrequest.png)

And see a history of your deployments in the **Environment** tab!

![activitylog](assets/activitylog.png)

## Usage

Install this package using `yarn` or `npm`:

```bash
yarn add -D github-deploy-status
```

```bash
npm i -D github-deploy-status
```

You can now use the command right in your shell!

```bash
github-deploy-status ...
```

Just make sure you give the correct arguments. See below for the arguments required.

## Arguments

#### `--token <token>`

This argument is _required_.

Interfacing with the Github API requires authentication. Create a
[personal API token](https://github.blog/2013-05-16-personal-api-tokens/)
and provide it to the script so that it can use it to authorise your
requests. Ensure that you grant your token the `repo_deployment` scope.

#### `--action <action>` or `-a <action>`

This argument is _required_.

`<action>` must be one of the following values:

- `create`
- `fail_if_unsuccessful` \*
- `success`
- `failure`
- `error`
- `in_progress`

Using `create` action will create a new deployment.

\* The `fail_if_unsuccessful` action is special - it will first **check** the status
of the latest deployment. If it is **NOT** successful (e.g. it is pending or in
progress), it will add a new `failure` status. This is useful to call at the end
of your deployment scripts in case the deployment doesn't complete and there is
no way to receive feedback other than to check the deployment status.

There must be an existing deployment before you add deployment statuses. The
remaining actions (`success`, `failure`, `error`, `in_progress`) will add the respective statuses
to the most recent deployment.

#### `--environment <environment>` or `-e <environment>`

This argument is _required_.

Use this to provide the name of the environment you are deploying to. Common
values for this are `dev`, `staging` or `production`.

#### `--ref <ref>` or `-r <ref>`

This argument is _required_.

Provide the ref of what you want to deploy from your repository. This can be
a tag, branch, or SHA.

#### `--user <username>` or `-u <username>`

This argument is _required_.

Provide the username of the user who owns the repository that you are
managing deployments for.

#### `--repo <repo_name>` or `-p <repo_name>`

This argument is _required_.

Provide the name of the repository that you are managing deployments for.

#### `--url <url>` or `-l <url>`

This argument is _optional_.

Provide a link to the deployed environment if you want the **View deployment**
button to show up next to your active deployments.

## Example

One way to take advantage of this script is to add it to your `package.json` and `.travis.yml`:

`package.json`:

```json
{
  ...
  "scripts": {
    "install-all": "yarn",
    "build-all": "tsc",
    "gds": "github-deploy-status -u a-user -p some-repository"
  }
  ...
}

```

`.travis.yml`:

```yml
...

jobs:
  include:
    - stage: Staging
      before_deploy:
        - yarn add github-deploy-status
        - yarn gds --token $GITHUB_TOKEN -a create -e staging -r develop
        - yarn gds --token $GITHUB_TOKEN -a in_progress -e staging -r develop
      deploy:
        ...
        on:
          branch: develop
      after_deploy:
        - yarn gds --token $GITHUB_TOKEN -a success -e staging -r develop -l https://staging.myapp.com

    - stage: Production
      before_deploy:
        - yarn add github-deploy-status
        - yarn gds --token $GITHUB_TOKEN -a create -e production -r master
        - yarn gds --token $GITHUB_TOKEN -a in_progress -e production -r master
      deploy:
        ...
        on:
          branch: master
      after_deploy:
        - yarn gds --token $GITHUB_TOKEN -a success -e production -r master -l https://myapp.com
...
```

In the above example, a deployment is created when Travis starts a deployment
(whose status is set to `pending` by default), and the status is set to `success`
once the deployment completes. The user's Github access token is stored as an
environment variable on Travis CI and is access through `$GITHUB_TOKEN`.

## License

[MIT License](LICENSE)


CHANGELOG.md
## [1.4.1](https://github.com/bwyap/github-deploy-status/compare/v1.4.0...v1.4.1) (2019-09-21)


### Features

* add logging of state ([1257f0a](https://github.com/bwyap/github-deploy-status/commit/1257f0a))



# [1.4.0](https://github.com/bwyap/github-deploy-status/compare/v1.3.2...v1.4.0) (2019-09-20)


### Bug Fixes

* in_progress no longer calls success afterwards ([6bdf742](https://github.com/bwyap/github-deploy-status/commit/6bdf742))


### Features

* add more informative logging ([963eac5](https://github.com/bwyap/github-deploy-status/commit/963eac5))



## [1.3.2](https://github.com/bwyap/github-deploy-status/compare/v1.3.1...v1.3.2) (2019-09-20)


### Bug Fixes

* make sure build is published ([a70d3ab](https://github.com/bwyap/github-deploy-status/commit/a70d3ab))



## [1.3.1](https://github.com/bwyap/github-deploy-status/compare/v1.3.0...v1.3.1) (2019-09-20)


### Bug Fixes

* fail_if_unsuccessful now runs if there are no deployments ([0c33166](https://github.com/bwyap/github-deploy-status/commit/0c33166))



# [1.3.0](https://github.com/bwyap/github-deploy-status/compare/v1.2.1...v1.3.0) (2019-09-20)


### Features

* add fail_if_unsuccessful action ([b1c62d1](https://github.com/bwyap/github-deploy-status/commit/b1c62d1))



## [1.2.1](https://github.com/bwyap/github-deploy-status/compare/v1.2.0...v1.2.1) (2019-09-19)


### Features

* add in_progress action ([43b97b8](https://github.com/bwyap/github-deploy-status/commit/43b97b8))



# [1.2.0](https://github.com/bwyap/github-deploy-status/compare/v1.1.0...v1.2.0) (2019-09-19)



# [1.1.0](https://github.com/bwyap/github-deploy-status/compare/v1.0.2...v1.1.0) (2019-09-19)


### Features

* make Node.js script into executable ([c843bdf](https://github.com/bwyap/github-deploy-status/commit/c843bdf))



## [1.0.2](https://github.com/bwyap/github-deploy-status/compare/v1.0.1...v1.0.2) (2019-09-19)



## [1.0.1](https://github.com/bwyap/github-deploy-status/compare/0fd1916...v1.0.1) (2019-09-19)


### Features

* create deployment update scripts ([0fd1916](https://github.com/bwyap/github-deploy-status/commit/0fd1916))





src/api.ts
import axios from "axios";

import { DeploymentStatus, Deployment, DeploymentState, Arguments } from "./types";
import { githubAPI } from "./constants";

const getDefaultHeaders = (token: string) => ({
  Authorization: `token ${token}`,
  "Content-Type": "application/json"
});

//
// Functions to make the required API calls to the Github API.
//
// REFERENCE
// https://developer.github.com/v3/repos/deployments
//

/**
 * Get a list of deployments for the specified repository.
 */
export const getDeployments = ({
  user,
  repo,
  token
}: Arguments) =>
  axios.get<Array<Deployment>>(
    `${githubAPI}/repos/${user}/${repo}/deployments`,
    {
      headers: getDefaultHeaders(token)
    }
  );

/**
 * Get a lsit of deployment statuses for the specified deployment.
 */
export const getDeploymentStatuses = (
  { user, repo, token }: Arguments,
  deployment: string
) =>
  axios.get<Array<DeploymentStatus>>(
    `${githubAPI}/repos/${user}/${repo}/deployments/${deployment}/statuses`,
    {
      headers: getDefaultHeaders(token)
    }
  );

/**
 * Create a deployment for the specified repository.
 */
export const createDeployment = ({
  user,
  repo,
  ref,
  environment,
  token
}: Arguments) =>
  axios.post(
    `${githubAPI}/repos/${user}/${repo}/deployments`,
    {
      ref,
      environment,
      auto_merge: false,
      required_contexts: []
    },
    {
      headers: {
        Accept: "application/vnd.github.v3+json",
        ...getDefaultHeaders(token)
      }
    }
  );

/**
 * Create a deployment status for the latest deployment in the specified repository.
 */
export const createDeploymentStatus = (
  {
    user,
    repo,
    environment,
    token,
    url
  }: Arguments,
  deployment: string,
  state: DeploymentState
) =>
  axios.post(
    `${githubAPI}/repos/${user}/${repo}/deployments/${deployment}/statuses`,
    {
      state,
      environment,
      environment_url: url
    },
    {
      headers: {
        Accept:
          "application/vnd.github.v3+json; application/vnd.github.flash-preview+json; application/vnd.github.ant-man-preview+json",
        ...getDefaultHeaders(token)
      }
    }
  );


src/actions.ts
import { Arguments, DeploymentState } from "./types";
import {
  createDeployment,
  createDeploymentStatus,
  getDeployments,
  getDeploymentStatuses
} from "./api";

/**
 * Create a new deployment for a ref in a repository.
 */
export const create = async (args: Arguments) => {
  console.log(
    `create: creating a deployment [${args.environment}] for ${args.user}/${args.repo} on <${args.ref}>...`
  );
  try {
    // Create deployment
    await createDeployment(args);

    console.log(`create: deployment created.`);
  } catch (error) {
    console.error(`create: ${error.toString()}`);
  }
};

/**
 * Helper function used to add a new deployment status.
 */
const addStatus = async (state: DeploymentState, args: Arguments) => {
  console.log(`addStatus: adding status [${state}] to latest deployment...`);

  try {
    // Get deployments
    const { data: deployments } = await getDeployments(args);

    if (deployments && deployments.length) {
      console.log(`addStatus: latest deployment found: ${deployments[0].id}`);
      console.log(`addStatus: creating "${state}" deployment status...`);

      // Create deployment status
      await createDeploymentStatus(args, deployments[0].id.toString(), state);

      console.log(
        `addStatus: deployment status for ${deployments[0].id} created.`
      );
    } else throw new Error(`addStatus: no deployments`);
  } catch (error) {
    console.error(`addStatus: ${error.toString()}`);
  }
};

/**
 * Add a success status to the last deployment.
 */
export const success = (args: Arguments) => addStatus("success", args);

/**
 * Add a in_progress status to the last deployment.
 */
export const progress = (args: Arguments) => addStatus("in_progress", args);

/**
 * Add an error status to the last deployment.
 */
export const error = (args: Arguments) => addStatus("error", args);

/**
 * Add a failure status to the last deployment.
 */
export const failure = (args: Arguments) => addStatus("failure", args);

/**
 * Check the latest deployment status. Add a `failure` status
 * if it is not one of `[successful, failure, error]`.
 */
export const failIfUnsuccessful = async (args: Arguments) => {
  console.log(
    `failIfUnsuccessful: adding failure state if deployment was unsuccessful...`
  );

  try {
    // Get the current deployments
    const { data: deployments } = await getDeployments(args);
    if (deployments && deployments.length) {
      console.log(
        `failIfUnsuccessful: latest deployment found: ${deployments[0].id}`
      );

      // Get the deployment statuses of the latest deployment
      const { data: statuses } = await getDeploymentStatuses(
        args,
        deployments[0].id.toString()
      );

      if (statuses && statuses.length) {
        console.log(
          `failIfUnsuccessful: latest deployment status found: ${statuses[0].id} [${statuses[0].state}]`
        );

        if (
          // Check the state of the latest deployment status
          statuses[0].state === "success" ||
          statuses[0].state === "inactive" ||
          statuses[0].state === "failure" ||
          statuses[0].state === "error"
        ) {
          // We can just return if the deployment status was successful or inactive.
          console.log(
            `failIfUnsuccessful: current state does not need modification.`
          );
          return;
        }
      }

      // Otherwise, set a failure status.
      console.log(
        `failIfUnsuccessful: setting failure status for deployment ${deployments[0].id}...`
      );
      await addStatus("failure", args);

      console.log(
        `failIfUnsuccessful: deployment status for ${deployments[0].id} updated.`
      );
    } else throw new Error(`failIfUnsuccessful: no deployments`);
  } catch (error) {
    console.error(`failIfUnsuccessful: ${error.toString()}`);
  }
};


src/types.d.ts
/**
 * The action to take to udpate the deployment status.
 *
 * `create` creates a new deployment.
 *
 * `in_progress`, `success`, `failure` and `error` add
 * a new deployment status to an existing deployment.
 *
 * `fail_if_in_progress` checks the current deployment
 * status, and if it is not `success`, add a `failure`
 * status.
 *
 */
export type Action =
  | "create"
  | "success"
  | "in_progress"
  | "failure"
  | "error"
  | "fail_if_unsuccessful";

/**
 * Possible deployment statuses.
 */
export type DeploymentState =
  | "error"
  | "failure"
  | "inactive"
  | "in_progress"
  | "queued"
  | "pending"
  | "success";

/**
 * The input arguments.
 */
export interface Arguments {
  /**
   * The action to update the deployment status.
   */
  action: Action;
  /**
   * The personal access token used to authenticate the
   * request to the Github API. See this link for more info:
   * https://github.blog/2013-05-16-personal-api-tokens/
   */
  token: string;
  /**
   * The environment the deployment status should be
   * placed under.
   */
  environment: string;
  /**
   * The username of the user who owns the repository
   * whose deployment status is being updated.
   */
  user: string;
  /**
   * The name of the repository whose deployment status
   * is being updated.
   */
  repo: string;
  /**
   * The name of the ref that the deployment is deploying
   * to.
   */
  ref: string;
  /**
   * The URL used to access the deployment.
   */
  url?: string;
}

/**
 * A deployment response from Github.
 */
export interface Deployment {
  url: string;
  id: number;
  node_id: string;
  sha: string;
  ref: string;
  task: string;
  payload: object;
  original_environment: string;
  environment: string;
  description: string | null;
  creator: {
    login: string;
    id: number;
    node_id: string;
    avatar_url: string;
    gravatar_id: string;
    url: string;
    html_url: string;
    followers_url: string;
    following_url: string;
    gists_url: string;
    starred_url: string;
    subscriptions_url: string;
    organizations_url: string;
    repos_url: string;
    events_url: string;
    received_events_url: string;
    type: string;
    site_admin: boolean;
  };
  created_at: string;
  updated_at: string;
  statuses_url: string;
  repository_url: string;
}

/**
 * A deployment status response from Github.
 */
export interface DeploymentStatus {
  url: string;
  id: number;
  node_id: string;
  state: DeploymentState;
  creator: {
    login: string;
    id: number;
    node_id: string;
    avatar_url: string;
    gravatar_id: string;
    url: string;
    html_url: string;
    followers_url: string;
    following_url: string;
    gists_url: string;
    starred_url: string;
    subscriptions_url: string;
    organizations_url: string;
    repos_url: string;
    events_url: string;
    received_events_url: string;
    type: string;
    site_admin: boolean;
  };
  description: string;
  environment: string;
  target_url: string;
  created_at: string;
  updated_at: string;
  deployment_url: string;
  repository_url: string;
  environment_url: string;
  log_url: string;
}


src/parser.ts
import { ArgumentParser } from "argparse";

const parser = new ArgumentParser({
  version: "1.3.1",
  description: `Node.js scripts that update a repository's Github deployment status.`,
  addHelp: true
});

parser.addArgument(["--action", "-a"], {
  optionStrings: [
    "create",
    "success",
    "in_progress",
    "failure",
    "error",
    "fail_if_unsuccessful"
  ],
  help: "The action to take on the deployment status.",
  required: true
});

parser.addArgument(["--token"], {
  type: "string",
  help:
    "Please provide a personal API token to authorise your requests. See this URL for more information: https://github.blog/2013-05-16-personal-api-tokens/",
  required: true
});

parser.addArgument(["--environment", "-e"], {
  type: "string",
  help:
    "Provide the name of the environment to apply the deployment status update to.",
  required: true
});

parser.addArgument(["--ref", "-r"], {
  type: "string",
  help: "Provide the ref that deployment status should be attached to.",
  required: true
});

parser.addArgument(["--user", "-u"], {
  type: "string",
  help: "Provide the username of the user who owns the repository.",
  required: true
});

parser.addArgument(["--repo", "-p"], {
  type: "string",
  help: "Provide the name the repository.",
  required: true
});

parser.addArgument(["--url", "-l"], {
  type: "string",
  help: "Provide the environment url for the deployment."
});

export default parser;
